File: leetcode/1282.用户分组.py
#
# @lc app=leetcode.cn id=1282 lang=python3
#
# [1282] 用户分组
#
from collections import defaultdict
from typing import List


# @lc code=start
class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        cahce_dict = defaultdict(list)
        for i in range(len(groupSizes)):
            cahce_dict[groupSizes[i]].append(i)
        return [
            peoples[i:i+groupSize]
            for groupSize, peoples in cahce_dict.items()
            for i in range(0, len(peoples), groupSize)
        ]


# @lc code=end
print(Solution().groupThePeople([3, 3, 3, 3, 3, 1, 3]))
print(Solution().groupThePeople([2, 1, 3, 3, 3, 2]))

File: leetcode/1282.用户分组.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;
/*
 * @lc app=leetcode.cn id=1282 lang=cpp
 *
 * [1282] 用户分组
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> groupThePeople(vector<int> &groupSizes)
    {
        unordered_map<int, vector<int>> cache;
        for (size_t i = 0; i < groupSizes.size(); i++)
        {
            cache[groupSizes[i]].emplace_back(i);
        }
        vector<vector<int>> res;
        for (auto &&c : cache)
        {
            int count = c.second.size() / c.first;
            for (size_t i = 0; i < count; i++)
            {
                vector<int> group;
                for (size_t j = 0; j < c.first; j++)
                {
                    group.emplace_back(c.second[i * c.first + j]);
                }
                res.emplace_back(group);
            }
        }
        return res;
    }
};
// @lc code=end
int main(int argc, char const *argv[])
{
    Solution s;
    int arr[] = {3, 3, 3, 3, 3, 1, 3};
    int n = sizeof(arr) / sizeof(arr[0]);
    vector<int> vect(arr, arr + n);
    vector<vector<int>> res = s.groupThePeople(vect);
    cout << "[";
    for (auto &&v : res)
    {
        cout << "[";
        for (auto &&i : v)
        {
            cout << i << ",";
        }
        cout << "]";
    }
    cout << "]";
    cout << endl;
    return 0;
}
